networkx/algorithms/approximation/treewidth.py

```diff
@@ -0,0 +1,272 @@
+                raise StopIteration
+
+            # Remember to update nodes in the heap before getting the next node
+            self._update_nodes = self._graph[elim_node]
+            return elim_node
+
+        # The heap is empty: Abort
+        raise StopIteration
+
+
+class MinFillInHeuristic:
+
+    def __init__(self, graph):
+        self._graph = graph
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        #Implement next method for backwards compatibility with python 2.
+        return self.__next__()
+
+    def __next__(self):
+        if len(self._graph) == 0:
+            raise StopIteration
+
+        min_fill_in_node = None
+
+        min_fill_in = sys.maxsize
+
+        # create sorted list of (degree, node)
+        degree_list = [(len(self._graph[node]), node) for node in self._graph]
+        degree_list.sort()
+
+        # abort condition
+        min_degree = degree_list[0][0]
+        if min_degree == len(self._graph) - 1:
+            raise StopIteration
+
+        for (_, node) in degree_list:
+            num_fill_in = 0
+            # Convert to list in order to access by index
+            nbrs = list(self._graph[node])
+            for i in range(len(nbrs) - 1):
+                for j in range(i + 1, len(nbrs)):
+                    if nbrs[j] not in self._graph[nbrs[i]]:
+                        num_fill_in += 1
+                        # break inner loop if this can't be min-fill-in node anymore
+                        if num_fill_in >= min_fill_in:
+                            break
+
+                if num_fill_in >= min_fill_in: # break outer loop
+                    break
+
+            if num_fill_in < min_fill_in: # Update min-fill-in node
+                if num_fill_in == 0:
+                    return node
+                min_fill_in = num_fill_in
+                min_fill_in_node = node
+
+        return min_fill_in_node
+
+
+def treewidth_decomp(G, heuristic_class):
+    """Returns a treewidth decomposition using the passed heuristic.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+    heuristic_class : iterator class
+
+    Returns
+    -------
+    Treewidth decomposition : (int, Graph) tuple
+        2-tuple with treewidth and the corresponding decomposed tree (NetworkX graph).
+    """
+    
+    # make dict-of-sets structure
+    graph = {}
+    for u in G:
+        graph[u] = set()
+        for v in G[u]:
+            if u != v:  # ignore self-loop
+                graph[u].add(v)
+
+    # stack where nodes and their neighbors are pushed in the order they are selected by the heuristic
+    node_stack = []
+
+    # instantiate a heuristic_iterator
+    heuristic_iterator = heuristic_class(graph)
+
+    for elim_node in heuristic_iterator:
+        # Connect all neighbours with each other
+        nbrs = graph[elim_node]
+        for u in nbrs:
+            for v in nbrs:
+                if u != v and v not in graph[u]:
+                    graph[u].add(v)
+
```

Comment:
Consider ```for u, v in itertools.combinations(nbrs, 2):``` and maybe rely on self-loops already being eliminated (if they really are and if you document that trait -- or check them here, but don't bother to eliminate them earlier).